src/laplace.jl

```diff
@@ -0,0 +1,336 @@
+    @debug "Hit rrule"
+    f_opt, cache = _newton_inner_loop(dist_y_given_f, ys, K; kwargs...)
+
+    # f = K (∇log p(y|f))                               (RW 3.17)
+    # δf = δK (∇log p(y|f)) + K δ(∇log p(y|f))
+    #    = δK (∇log p(y|f)) + K ∇(∇log p(y|f)) δf
+    # δf (I - K ∇∇log p(y|f)) = δK (∇log p(y|f))
+    # δf (I + K W) = δK (∇log p(y|f))
+    # δf = (I + K W)⁻¹ δK (∇log p(y|f))                 (RW 5.24)
+    # (I + K W) = (√W)⁻¹ (I + √W K √W) √W = (√W)⁻¹ B √W
+    # δf = (√W)⁻¹ B⁻¹ √W δK (∇log p(y|f))
+
+    # ∂f_opt = cache.Wsqrt \ (cache.B_ch \ (cache.Wsqrt * (ΔK * cache.d_loglik)))
+
+    # Re<Δf, δf> = Re<Δf, Wsqrt\inv B\inv Wsqrt δK d_loglik>
+    #            = Re<Wsqrt' B\inv' Wsqrt\inv' Δf d_loglik', δK>
+    #
+    # ΔK = Wsqrt' * cache.B_ch' \ Wsqrt' \ Δf_opt * cache.d_loglik'
+
+    function newton_pullback(Δf_opt)
+        ∂self = NoTangent()
+
+        ∂dist_y_given_f = @not_implemented(
+            "gradient of Newton's method w.r.t. likelihood parameters"
+        )
+
+        ∂ys = @not_implemented("gradient of Newton's method w.r.t. observations")
+
+        # ∂K = df/dK Δf
+        ∂K = @thunk(cache.Wsqrt * (cache.B_ch \ (cache.Wsqrt \ Δf_opt)) * cache.d_loglik')
+
+        return (∂self, ∂dist_y_given_f, ∂ys, ∂K)
+    end
+
+    return f_opt, newton_pullback
+end
+
+function laplace_lml(dist_y_given_f, ys, K, f_opt)
+    cache = _laplace_train_intermediates(dist_y_given_f, ys, K, f_opt)
+    return _laplace_lml(f_opt, cache)
+end
+
+function laplace_lml(
+    dist_y_given_f, ys, K; f_init=zeros(length(ys)), maxiter=100, newton_kwargs...
+)
+    f_opt = newton_inner_loop(dist_y_given_f, ys, K; f_init, maxiter, newton_kwargs...)
+    return laplace_lml(dist_y_given_f, ys, K, f_opt)
+end
+
+function laplace_lml(lfx::LatentFiniteGP, ys; newton_kwargs...)
+    dist_y_given_f, K, newton_kwargs = _check_laplace_inputs(lfx, ys; newton_kwargs...)
+    return laplace_lml(dist_y_given_f, ys, K; newton_kwargs...)
+end
+
+function laplace_f_and_lml(lfx::LatentFiniteGP, ys; newton_kwargs...)
+    dist_y_given_f, K, newton_kwargs = _check_laplace_inputs(lfx, ys; newton_kwargs...)
+    f_opt = newton_inner_loop(dist_y_given_f, ys, K; newton_kwargs...)
+    lml = laplace_lml(dist_y_given_f, ys, K, f_opt)
+    return f_opt, lml
+end
+
+function _check_laplace_inputs(
+    lfx::LatentFiniteGP, ys; f_init=nothing, maxiter=100, newton_kwargs...
+)
+    fx = lfx.fx
+    @assert mean(fx) == zero(mean(fx))  # might work with non-zero prior mean but not checked
+    @assert length(ys) == length(fx)
+    dist_y_given_f = lfx.lik
+    K = cov(fx)
+    if isnothing(f_init)
+        f_init = mean(fx)
+    end
+    return dist_y_given_f, K, (; f_init, maxiter, newton_kwargs...)
+end
+
+function build_laplace_objective!(
+    f,
+    build_latent_gp,
+    xs,
+    ys;
+    newton_warmstart=true,
+    newton_callback=nothing,
+    newton_maxiter=100,
+)
+    initialize_f = true
```

Comment:
There seems to be some type instability around this, but as long as the final return value (`lml`) is inferred correctly, I am guessing it doesn't matter ?